symengine/sets.cpp

```diff
@@ -301,4 +317,159 @@ const RCP<const UniversalSet> &UniversalSet::getInstance()
     const static auto a = make_rcp<const UniversalSet>();
     return a;
 }
+
+FiniteSet::FiniteSet(const set_basic container) : container_(container)
+{
+    SYMENGINE_ASSERT(FiniteSet::is_canonical(container_));
+}
+
+bool FiniteSet::is_canonical(const set_basic container)
+{
+    return container.size() != 0;
+}
+
+std::size_t FiniteSet::__hash__() const
+{
+    std::size_t seed = FINITESET;
+    for (const auto &a : container_)
+        hash_combine<Basic>(seed, *a);
+    return seed;
+}
+
+bool FiniteSet::__eq__(const Basic &o) const
+{
+    if (is_a<FiniteSet>(o)) {
+        const FiniteSet &other = static_cast<const FiniteSet &>(o);
+        if (container_.size() != other.container_.size())
+            return false;
+        for (const auto &a : container_) {
+            if (not other.contains(a))
+                return false;
+        }
+        // TODO : earch for a std function
+        return true;
+    }
+    if (is_a<Interval>(o)) {
+        // Interval and FiniteSet can be equal if Interval is only a point
+        const Interval &other = static_cast<const Interval &>(o);
```

Comment:
This is not going to work. You need to have hashes equal for both and the type code as well for some algorithms.

You can return a FiniteSet from an Interval constructor if it only a point